### PR TITLE
added error handling for invalid handlers and some debug messages

### PR DESCRIPTION
**Pull request by cap5lut**
Hi,
first of all, sorry if I did something wrong, its the first time I fork/create a pull request.
Regarding the issue #5 I thought of some tiny additions.

I added some debug messages for example if the controller that gets passed is not an object (basically what caused my problems regarding issue #5,
wrote a little check, if those handlers are actually existing before they are called,
and made it configurable if an exception shall be thrown if they can not be called.

cap5lut

### CODE REVIEW

**Comment by AlessandroMinoccheri** [comment]
No problem, there is a first time for all :)

But can you please move the class to manage exception into another file?

After can you please add tests for that piece of code?
Is possible?

Thanks for your contribute!

P.S. remember that I have make some little edit, so make a rebase of your branch with new version 3.0.2 to be sure if you have take code from v 3.0.1

**Comment by cap5lut** [comment]
I'll do it as soon as possible, currently I'm busy.

**Comment by AlessandroMinoccheri** [comment]
Ok I'll waiting for your update :)

**Comment by cap5lut** [comment]
Sorry for the long delay, moved the exception to its own file and added the tests, the base is at v3.0.2 as well

File: src/Exception/MissingHandlerException.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+use Cake\Core\Exception\Exception;
+
+/**
+ * An instance of this exception should be thrown, if the
+ * UserPermissionsComponent instance tries to call an handler which does not
+ * exist.
+ */
+class MissingHandlerException extends Exception
+{};
```
Comment (AlessandroMinoccheri) [inline src/Exception/MissingHandlerException.php]: add a new line at the end of the file

File: tests/TestCase/Controller/Component/UserPermissionComponentTest.php [hunk 2]
```diff
@@ -176,4 +177,59 @@ public function testUserWithPermissionAndTrueCallback() {
...
+			$this->fail("No exception thrown");
+		}
+		catch(MissingHandlerException $e) {
+			
+		}
+		catch(\Exception $e) {
+			$this->fail("Wrong exception thrown " . get_class($e));
+		}
+	}
 }
```
Comment (AlessandroMinoccheri) [inline tests/TestCase/Controller/Component/UserPermissionComponentTest.php]: add a new line at the end of the file

File: tests/TestCase/Controller/Component/UserPermissionComponentTest.php [hunk 3]
```diff
@@ -176,4 +177,59 @@ public function testUserWithPermissionAndTrueCallback() {
...
+		
+		try {
+			$result = $this->userPermissions->allow($rules);
+			$this->fail("No exception thrown");
+		}
+		catch(MissingHandlerException $e) {
+			
+		}
+		catch(\Exception $e) {
+			$this->fail("Wrong exception thrown " . get_class($e));
```
Comment (AlessandroMinoccheri) [inline tests/TestCase/Controller/Component/UserPermissionComponentTest.php]: In this case is better to use annotation for example read this:

https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.expectedException

File: tests/TestCase/Controller/Component/UserPermissionComponentTest.php (hunk 2 above)
Comment (cap5lut) [reply tests/TestCase/Controller/Component/UserPermissionComponentTest.php]: done

**Pull request merged**